File: packages/ui/src/standards/symbol-generators/accessibility.ts
import type { AccessibilityElement } from '../../services/kernel-bridge'
import type { SymbolGenerationOptions, SymbolPrimitiveBundle } from '../symbol-primitives'
import { getDefaultSymbolProvenance, getSymbolSpec } from '../plan-symbol-profile'
import {
  addRect, addLine, addCircle, addText,
  type SymbolPrimitive,
} from './_helpers'

function addWheelchairSymbol(
  primitives: SymbolPrimitive[],
  cx: number,
  cy: number,
  rotation: number,
  scale: number,
): void {
  // Simplified ISA wheelchair icon
  // Head
  addCircle(primitives, [cx, cy], rotation, [0, 0.14 * scale], 0.03 * scale, 'primary')
  // Body vertical line
  addLine(primitives, [cx, cy], rotation, [0, 0.11 * scale], [0, 0.02 * scale], 'primary')
  // Arm line
  addLine(primitives, [cx, cy], rotation, [0, 0.08 * scale], [0.06 * scale, 0.06 * scale], 'primary')
  // Large wheel
  addCircle(primitives, [cx, cy], rotation, [0, -0.02 * scale], 0.07 * scale, 'primary')
  // Small front wheel
  addCircle(primitives, [cx, cy], rotation, [0.06 * scale, -0.02 * scale], 0.02 * scale, 'primary')
  // Footrest
  addLine(primitives, [cx, cy], rotation, [0.03 * scale, -0.02 * scale], [0.08 * scale, -0.06 * scale], 'primary')
}

export function generateAccessibilityPrimitives(
  element: AccessibilityElement,
  options: SymbolGenerationOptions,
): SymbolPrimitiveBundle {
  const spec = getSymbolSpec(options.planSymbolProfile, 'accessibility', element.symbol_type) ?? {
    domain: 'accessibility' as const,
    title: element.symbol_type,
    lineClass: 'primary' as const,
    allowsLabel: false,
    provenance: getDefaultSymbolProvenance(),
  }
  const primitives: SymbolPrimitive[] = []
  const [x, y] = element.position
  const rotation = element.rotation ?? 0

  switch (element.symbol_type) {
    case 'wheelchair': {
      addWheelchairSymbol(primitives, x, y, rotation, 1.0)
      break
    }

    case 'ramp': {
      // Two parallel lines
      addLine(primitives, [x, y], rotation, [-0.30, 0.075], [0.30, 0.075], 'primary')
      addLine(primitives, [x, y], rotation, [-0.30, -0.075], [0.30, -0.075], 'primary')
      // Direction arrow
      addLine(primitives, [x, y], rotation, [0.20, 0], [0.30, 0], 'secondary')
      addLine(primitives, [x, y], rotation, [0.26, 0.03], [0.30, 0], 'secondary')
      addLine(primitives, [x, y], rotation, [0.26, -0.03], [0.30, 0], 'secondary')
      // Slope label
      if (options.showLabels || options.annotationDensity !== 'minimal') {
        addText(primitives, [x, y], rotation, '1:12', 0.06)
      }
      break
    }

    case 'grab_bar': {
      // Heavy line for the bar
      addLine(primitives, [x, y], rotation, [-0.30, 0], [0.30, 0], 'cut')
      // Mounting dots at each end
      addCircle(primitives, [x, y], rotation, [-0.30, 0], 0.025, 'primary')
      addCircle(primitives, [x, y], rotation, [0.30, 0], 0.025, 'primary')
      break
    }

    case 'accessible_parking': {
      // Parking space rectangle
      addRect(primitives, [x, y], rotation, 1.0, 0.50, 'primary')
      // Scaled wheelchair symbol inside
      addWheelchairSymbol(primitives, x, y, rotation, 0.6)
      break
    }

    case 'tactile_warning': {
      // Rectangle outline
      addRect(primitives, [x, y], rotation, 0.40, 0.30, 'primary')
      // Grid of small dots (4 columns x 3 rows)
      for (let row = 0; row < 3; row++) {
        for (let col = 0; col < 4; col++) {
          const lx = -0.135 + col * 0.09
          const ly = -0.09 + row * 0.09
          addCircle(primitives, [x, y], rotation, [lx, ly], 0.012, 'secondary')
        }
      }
      break
    }

    case 'ada_restroom': {
      addRect(primitives, [x, y], rotation, 1.50, 1.50, 'primary')
      addText(primitives, [x, y], rotation, 'ADA', 0.12)
      addWheelchairSymbol(primitives, x, y, rotation, 0.5)
      break
    }

    case 'hearing_loop': {
      addCircle(primitives, [x, y], rotation, [0, 0], 0.15, 'primary')
      addText(primitives, [x, y], rotation, 'HL', 0.08)
      break
    }

    default: {
      addCircle(primitives, [x, y], rotation, [0, 0], 0.15, spec.lineClass as 'primary')
      break
    }
  }

  return {
    domain: spec.domain,
    primitives,
  }
}

File: crates/bcad-2d/src/symbols/accessibility.rs
//! Accessibility symbol generators.
//!
//! Geometry closely mirrors the TypeScript generators in
//! `packages/ui/src/standards/symbol-generators/accessibility.ts`.

use crate::style::{DomainColor, StyledPrimitive, SymbolLineClass};

use super::{add_circle, add_line, add_rect, add_text};

const DOMAIN: DomainColor = DomainColor::Accessibility;

/// Helper: simplified ISA wheelchair icon at a given center and scale.
fn add_wheelchair_symbol(
    out: &mut Vec<StyledPrimitive>,
    cx: f64,
    cy: f64,
    rotation: f64,
    scale: f64,
) {
    // Head
    add_circle(
        out,
        DOMAIN,
        [cx, cy],
        rotation,
        [0.0, 0.14 * scale],
        0.03 * scale,
        SymbolLineClass::Primary,
        false,
    );
    // Body vertical line
    add_line(
        out,
        DOMAIN,
        [cx, cy],
        rotation,
        [0.0, 0.11 * scale],
        [0.0, 0.02 * scale],
        SymbolLineClass::Primary,
        false,
    );
    // Arm line
    add_line(
        out,
        DOMAIN,
        [cx, cy],
        rotation,
        [0.0, 0.08 * scale],
        [0.06 * scale, 0.06 * scale],
        SymbolLineClass::Primary,
        false,
    );
    // Large wheel
    add_circle(
        out,
        DOMAIN,
        [cx, cy],
        rotation,
        [0.0, -0.02 * scale],
        0.07 * scale,
        SymbolLineClass::Primary,
        false,
    );
    // Small front wheel
    add_circle(
        out,
        DOMAIN,
        [cx, cy],
        rotation,
        [0.06 * scale, -0.02 * scale],
        0.02 * scale,
        SymbolLineClass::Primary,
        false,
    );
    // Footrest
    add_line(
        out,
        DOMAIN,
        [cx, cy],
        rotation,
        [0.03 * scale, -0.02 * scale],
        [0.08 * scale, -0.06 * scale],
        SymbolLineClass::Primary,
        false,
    );
}

/// Wheelchair accessibility symbol (ISA icon).
pub fn wheelchair(center: [f64; 2], rotation: f64) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    add_wheelchair_symbol(&mut out, center[0], center[1], rotation, 1.0);
    out
}

/// Ramp: two parallel lines + direction arrow + slope label.
pub fn ramp(center: [f64; 2], rotation: f64, show_label: bool) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    add_line(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [-0.30, 0.075],
        [0.30, 0.075],
        SymbolLineClass::Primary,
        false,
    );
    add_line(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [-0.30, -0.075],
        [0.30, -0.075],
        SymbolLineClass::Primary,
        false,
    );
    // Direction arrow
    add_line(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [0.20, 0.0],
        [0.30, 0.0],
        SymbolLineClass::Annotation,
        false,
    );
    add_line(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [0.26, 0.03],
        [0.30, 0.0],
        SymbolLineClass::Annotation,
        false,
    );
    add_line(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [0.26, -0.03],
        [0.30, 0.0],
        SymbolLineClass::Annotation,
        false,
    );
    if show_label {
        add_text(&mut out, DOMAIN, center, rotation, "1:12", 0.06);
    }
    out
}

/// Grab bar: heavy line + mounting dots at each end.
pub fn grab_bar(center: [f64; 2], rotation: f64) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    add_line(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [-0.30, 0.0],
        [0.30, 0.0],
        SymbolLineClass::Primary,
        false,
    );
    add_circle(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [-0.30, 0.0],
        0.025,
        SymbolLineClass::Primary,
        false,
    );
    add_circle(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [0.30, 0.0],
        0.025,
        SymbolLineClass::Primary,
        false,
    );
    out
}

/// Accessible parking: rectangle + scaled wheelchair icon.
pub fn accessible_parking(center: [f64; 2], rotation: f64) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    add_rect(
        &mut out,
        DOMAIN,
        center,
        rotation,
        1.0,
        0.50,
        SymbolLineClass::Primary,
        [0.0, 0.0],
        false,
    );
    add_wheelchair_symbol(&mut out, center[0], center[1], rotation, 0.6);
    out
}

/// Tactile warning: rectangle + grid of small dots.
pub fn tactile_warning(center: [f64; 2], rotation: f64) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    add_rect(
        &mut out,
        DOMAIN,
        center,
        rotation,
        0.40,
        0.30,
        SymbolLineClass::Primary,
        [0.0, 0.0],
        false,
    );
    for row in 0..3 {
        for col in 0..4 {
            let lx = -0.135 + col as f64 * 0.09;
            let ly = -0.09 + row as f64 * 0.09;
            add_circle(
                &mut out,
                DOMAIN,
                center,
                rotation,
                [lx, ly],
                0.012,
                SymbolLineClass::Secondary,
                false,
            );
        }
    }
    out
}

/// ADA restroom: large rectangle + "ADA" label + wheelchair icon.
pub fn ada_restroom(center: [f64; 2], rotation: f64) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    add_rect(
        &mut out,
        DOMAIN,
        center,
        rotation,
        1.50,
        1.50,
        SymbolLineClass::Primary,
        [0.0, 0.0],
        false,
    );
    add_text(&mut out, DOMAIN, center, rotation, "ADA", 0.12);
    add_wheelchair_symbol(&mut out, center[0], center[1], rotation, 0.5);
    out
}

/// Hearing loop: circle + "HL" label.
pub fn hearing_loop(center: [f64; 2], rotation: f64) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    add_circle(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [0.0, 0.0],
        0.15,
        SymbolLineClass::Primary,
        false,
    );
    add_text(&mut out, DOMAIN, center, rotation, "HL", 0.08);
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn wheelchair_produces_primitives() {
        let prims = wheelchair([0.0, 0.0], 0.0);
        assert!(
            prims.len() >= 6,
            "wheelchair icon should have head + body + arm + 2 wheels + footrest"
        );
    }

    #[test]
    fn tactile_warning_produces_primitives() {
        let prims = tactile_warning([0.0, 0.0], 0.0);
        // rect + 12 dots = 13
        assert!(prims.len() >= 13);
    }
}
